### PR TITLE
planner: fix uncorrect behavior of index join

### LINKED ISSUE #11390: wrong results of `IndexLookupJoin` in some cases
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
When building ranges for `IndexLookupJoin` on an expression like `Column=Constant`, it doesn't take their types into account ([at this line of code](https://github.com/pingcap/tidb/blob/master/planner/core/exhaust_physical_plans.go#L949)).
```
create table t (k1 int unsigned, k2 int unsigned, key(k1, k2));
insert into t values (1, 1);
select /*+ TIDB_INLJ(t1, t2) */ * from t t1, t t2 where t1.k2 > 0 and t1.k2 = t2.k2 and t2.k1=1;
select /*+ TIDB_HJ(t1, t2) */ * from t t1, t t2 where t1.k2 > 0 and t1.k2 = t2.k2 and t2.k1=1;
```

2. What did you expect to see?
These two queries should return the same result;

3. What did you see instead?
The query with `TIDB_INLJ` hint returns nothing.
```
mysql> select /*+ TIDB_INLJ(t1, t2) */ * from t t1, t t2 where t1.k2 > 0 and t1.k2 = t2.k2 and t2.k1=1;
Empty set (0.00 sec)

mysql> select /*+ TIDB_HJ(t1, t2) */ * from t t1, t t2 where t1.k2 > 0 and t1.k2 = t2.k2 and t2.k1=1;
+------+------+------+------+
| k1   | k2   | k1   | k2   |
+------+------+------+------+
|    1 |    1 |    1 |    1 |
+------+------+------+------+
1 row in set (0.00 sec)
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
```
mysql> select version();
+---------------------------------------+
| version()                             |
+---------------------------------------+
| 5.7.25-TiDB-list-313-gca70d74a2-dirty |
+---------------------------------------+
1 row in set (0.00 sec)

```

### LINKED ISSUE #11544: Index Join executes wrongly when there's index with column has length
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.

```
create table t(a int);
create table tt(a int, b varchar(10), index idx(a, b(3)));
insert into t values(1);
insert into tt values(1, 'aaaaaaa'), (1, 'aaaabbb'), (1, 'aaaacccc');
select /*+ TIDB_HJ(tt) */ * from t, tt where t.a=tt.a and tt.b in ('aaaaaaa', 'aaaabbb', 'aaaacccc');
select /*+ TIDB_INLJ(tt) */ * from t, tt where t.a=tt.a and tt.b in ('aaaaaaa', 'aaaabbb', 'aaaacccc');
```

2. What did you expect to see?

Both them returns correct dataset.

3. What did you see instead?

Only hash join returned. Index join returned a empty dataset.

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

close #11544 and close #11390 

### What is changed and how it works?

Use `BuildColumnRange` to build template range for index join.
To fix two cases:
1. The type of column and constant may be different.
2. The column may have length.

### Check List 

Tests 

 - Unit test

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: util/ranger/ranger_test.go [hunk 1]
```diff
@@ -1018,6 +1018,15 @@ func (s *testRangerSuite) TestColumnRange(c *C) {
 			resultStr:   "[[\"a\",\"a\"]]",
 			length:      1,
 		},
+		// This test case cannot be simplified to [1, 3] otherwise the index join will executes wrongly.
```
Comment (winoros) [inline util/ranger/ranger_test.go]: Use this test to assert that there won't be range merging in this case.

**Comment by lzmhhh123** [comment]
@winoros ci failed. Σ(° △ °|||)︴

**Comment by winoros** [comment]
@lzmhhh123 Yeah, i'm solving another one. Will fix it later.

**Comment by winoros** [comment]
@lzmhhh123  fixed

File: util/ranger/ranger_test.go (hunk 1 above)
Comment (lzmhhh123) [reply util/ranger/ranger_test.go]: Why can't merge to [1,3]? Because of the prefix index?

File: planner/core/exhaust_physical_plans.go [hunk 2]
```diff
@@ -448,7 +420,10 @@ func (p *LogicalJoin) getIndexJoinByOuterIdx(prop *property.PhysicalProperty, ou
 			continue
 		}
 		indexInfo := path.index
-		err := helper.analyzeLookUpFilters(indexInfo, ds, innerJoinKeys)
+		emptyRange, err := helper.analyzeLookUpFilters(indexInfo, ds, innerJoinKeys)
+		if emptyRange {
```
Comment (winoros) [inline planner/core/exhaust_physical_plans.go]: Otherwise we may need to return a `HashJoin{Reader, Dual}`. Which i think is a little strange.

Comment (lamxTyler) [reply planner/core/exhaust_physical_plans.go]: Can you add a unit test to cover it?

Comment (winoros) [reply planner/core/exhaust_physical_plans.go]: added

**Pull request merged**